src/v2/providers/storage.ts

```diff
@@ -0,0 +1,419 @@
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object finalized */
+export function onObjectFinalized(
+  bucket: string,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object finalized */
+export function onObjectFinalized(
+  opts: StorageOptions,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+export function onObjectFinalized(
+  optsOrHandler:
+    | string
+    | StorageOptions
+    | ((event: CloudEvent<StorageObjectData>) => any | Promise<any>),
+  handler?: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData> {
+  if (arguments.length === 1) {
+    handler = optsOrHandler as (
+      event: CloudEvent<StorageObjectData>
+    ) => any | Promise<any>;
+    _onOperation(handler, finalizedEvent);
+  }
+  return _onOperation(
+    handler,
+    finalizedEvent,
+    optsOrHandler as string | StorageOptions
+  );
+}
+
+/** Handle a storage object deleted */
+export function onObjectDeleted(
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object deleted */
+export function onObjectDeleted(
+  bucket: string,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object deleted */
+export function onObjectDeleted(
+  opts: StorageOptions,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+export function onObjectDeleted(
+  optsOrHandler:
+    | string
+    | StorageOptions
+    | ((event: CloudEvent<StorageObjectData>) => any | Promise<any>),
+  handler?: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData> {
+  if (arguments.length === 1) {
+    handler = optsOrHandler as (
+      event: CloudEvent<StorageObjectData>
+    ) => any | Promise<any>;
+    _onOperation(handler, deletedEvent);
+  }
+  return _onOperation(
+    handler,
+    deletedEvent,
+    optsOrHandler as string | StorageOptions
+  );
+}
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  bucket: string,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  opts: StorageOptions,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+export function onObjectMetadataUpdated(
+  optsOrHandler:
```

Comment:
nits* It's a mouthful, but I think I would prefer `butcketOrOptsOrHandler` for accuracy.

Reply:
Makes sense, fixed